spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/ObjectToKeyFactory.java

```diff
@@ -0,0 +1,93 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.datastore.core.convert;
+
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.KeyFactory;
+
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntity;
+import org.springframework.data.mapping.PersistentProperty;
+import org.springframework.util.Assert;
+
+/**
+ * An interface for creating Datastore Keys from objects and ID values.
+ *
+ * @author Chengyuan Zhao
+ *
+ * @since 1.1
+ */
+public interface ObjectToKeyFactory {
+
+	/**
+	 * Gets a key factory for creating new Keys.
+	 * @return a key factory.
+	 */
+	KeyFactory getKeyFactory();
```

Comment:
This interface is more like a template pattern implementation and not a true contract.
It's like a `KeyFactory` provider because all you need to implement is the `getKeyFactory` method.
Instead I think the `KeyFactory` should be kept out of the interface because it's more of an implementation detail.
So, I would make the default methods just interface methods, remove `getKeyFactory`, and then move the `default` methods to a default implementation of our key factory interface.

Reply:
done, PTAL